### PR TITLE
DNS and Host options

### PR DESCRIPTION
**Pull request by baude**
Add --dns-search, --dns-opt, --dns-server and --add-host.

Each of these options are destructive in nature, meaning if the user
adds one of them, all current ones are removed from the produced
resolv.conf.

* dns-server allows the user to specify dns servers.
* dns-opt allows the user to specify special resolv.conf options
* dns-search allows the user to specify search domains

The add-host option is not destructive and truly just adds the host
to /etc/hosts.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
@baude some of the network tests in the Red Hat CI aren't happy.

File: libpod/options.go [hunk 1]
```diff
@@ -502,3 +503,58 @@ func WithPodLabels(labels map[string]string) PodCreateOption {
...
 }
+
+// WithDNSSearch sets the additional search domains of a container
+func WithDNSSearch(searchDomains []string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return ErrCtrFinalized
+		}
+		if StringInSlice(".", searchDomains) && len(searchDomains) > 1 {
+			return errors.Errorf("cannot pass addition search domains when also specifying '.'")
```
Comment (TomSweeneyRedHat) [inline libpod/options.go]: nit: addition to additional

**Comment by baude** [comment]
nope it's all been there.

**Comment by baude** [comment]
[(3) [kpod-run] Implement kpod run CLI option: --dns=[]](https://trello.com/c/8RzZxYqW/492-3-kpod-run-implement-kpod-run-cli-option-dns)

**Comment by baude** [comment]
[(3) [kpod-run] Implement podman run CLI option: --dns-option=[]](https://trello.com/c/I9gv9gsT/493-3-kpod-run-implement-podman-run-cli-option-dns-option)

**Comment by baude** [comment]
[(3) [kpod-run] Implement podman run CLI option: --dns-search=[]](https://trello.com/c/nd5yUNSt/499-3-kpod-run-implement-podman-run-cli-option-dns-search)

**Comment by baude** [comment]
[(3) [kpod-run] Implement kpod run CLI option: --add-host=[]](https://trello.com/c/168D8h94/468-3-kpod-run-implement-kpod-run-cli-option-add-host)

File: libpod/options.go (hunk 1 above)
Comment (baude) [reply libpod/options.go]: fixed

**Comment by baude** [comment]
bot, restest this please

File: libpod/container.go [hunk 2]
```diff
@@ -34,6 +33,7 @@ import (
 	"golang.org/x/sys/unix"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/tools/remotecommand"
+	"strings"
```
[the remark this answers is not in the document]
Comment (baude) [reply libpod/container.go]: fixed

File: cmd/podman/create.go [hunk 3]
```diff
@@ -554,6 +555,14 @@ func parseCreateOpts(c *cli.Context, runtime *libpod.Runtime, imageName string,
 		networkMode = c.String("net")
 	}
 
+	// Verify the additional hosts are in correct format
+	for _, host := range c.StringSlice("add-host") {
+		_, err := validateExtraHost(host)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (baude) [reply cmd/podman/create.go]: fixed

File: libpod/container.go [hunk 4]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+
+// generateResolvConf generates a containers resolv.conf
+func (c *Container) generateResolvConf() (string, error) {
+	// Copy /etc/resolv.conf to the container's rundir
+	resolvPath := "/etc/resolv.conf"
+
+	// Check if the host system is using system resolve and if so
+	// copy its resolv.conf
+	_, err := os.Stat("/run/systemd/resolve/resolv.conf")
+	if err == nil {
```
Comment (rhatdan) [inline libpod/container.go]: Nit One line
if _, err := ...

File: libpod/options.go [hunk 5]
```diff
@@ -636,3 +637,58 @@ func WithPodLabels(labels map[string]string) PodCreateOption {
...
 	}
 }
+
+// WithDNSSearch sets the additional search domains of a container
+func WithDNSSearch(searchDomains []string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return ErrCtrFinalized
+		}
+		if StringInSlice(".", searchDomains) && len(searchDomains) > 1 {
```
Comment (mheon) [inline libpod/options.go]: Can you do this check in podman and make this a straight assignment? An empty []string can mean no search domains.

File: libpod/container.go [hunk 6]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+	orig, err := ioutil.ReadFile(resolvPath)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to read %s", resolvPath)
+	}
+	if len(c.config.DNSServer) == 0 && len(c.config.DNSSearch) == 0 && len(c.config.DNSOption) == 0 {
+		return c.WriteStringToRundir("resolv.conf", fmt.Sprintf("%s", orig))
+	}
+
+	// Read and organize the hosts /etc/resolv.conf
+	resolv := createResolv(string(orig[:]))
```
Comment (rhatdan) [inline libpod/container.go]: Is the ":" necessary here?

File: libpod/options.go [hunk 7]
```diff
@@ -636,3 +637,58 @@ func WithPodLabels(labels map[string]string) PodCreateOption {
...
+}
+
+// WithDNS sets additional name servers for the container
+func WithDNS(dnsServers []string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return ErrCtrFinalized
+		}
+		var dns []net.IP
+		for _, i := range dnsServers {
```
Comment (mheon) [inline libpod/options.go]: Should we check to make sure they gave us at least 1 DNS server here? Are there cases where having no DNS servers is something you want?

File: libpod/container.go [hunk 8]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+
+	// Populate the resolv struct with user's dns search domains
+	if len(c.config.DNSSearch) > 0 {
+		resolv.searchDomains = nil
+		// The . character means the user doesnt want any search domains in the container
+		if !StringInSlice(".", c.config.DNSSearch) {
+			resolv.searchDomains = append(resolv.searchDomains, c.Config().DNSSearch...)
+		}
```
Comment (rhatdan) [inline libpod/container.go]: Should we remove the search domains if the user specified ".", if one did exist in the resolv.conf from the host?

File: libpod/container.go [hunk 9]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+	}
+
+	// Populate the resolv struct with user's dns servers
+	if len(c.config.DNSServer) > 0 {
```
Comment (rhatdan) [inline libpod/container.go]: No need for length check

File: libpod/container.go [hunk 10]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+		resolv.nameServers = nil
+		for _, i := range c.config.DNSServer {
+			resolv.nameServers = append(resolv.nameServers, i.String())
+		}
+	}
+
+	// Populate the resolve struct with the users dns options
+	if len(c.config.DNSOption) > 0 {
+		resolv.options = nil
```
Comment (rhatdan) [inline libpod/container.go]: Is there any need for this?

File: libpod/container.go [hunk 11]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+
+//ToString returns a resolv struct in the form of a resolv.conf
+func (r resolv) ToString() string {
+	var result string
+	// Populate the output string with search domains
+	if len(r.searchDomains) > 0 {
+		result += fmt.Sprintf("search %s\n", strings.Join(r.searchDomains, " "))
+	}
+	// Populate the output string with name servers
+	if len(r.nameServers) > 0 {
```
Comment (rhatdan) [inline libpod/container.go]: No need for this check

File: libpod/container.go [hunk 12]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+		for _, i := range r.nameServers {
+			result += fmt.Sprintf("nameserver %s\n", i)
+		}
+	}
+	// Populate the output string with dns options
+	if len(r.options) > 0 {
```
Comment (rhatdan) [inline libpod/container.go]: No need for this check

File: libpod/container.go [hunk 13]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+}
+
+// generateHosts creates a containers hosts file
+func (c *Container) generateHosts() (string, error) {
+	orig, err := ioutil.ReadFile("/etc/hosts")
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to read /etc/hosts")
+	}
+	hosts := string(orig)
+	if len(c.config.HostAdd) > 0 {
```
Comment (rhatdan) [inline libpod/container.go]: No need for this check

File: libpod/options.go [hunk 14]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/containers/storage/pkg/idtools"
 	"github.com/cri-o/ocicni/pkg/ocicni"
 	"github.com/pkg/errors"
+	"net"
```
Comment (rhatdan) [inline libpod/options.go]: Add builtin golang to the top not to the github section.

File: libpod/options.go [hunk 15]
```diff
@@ -636,3 +637,58 @@ func WithPodLabels(labels map[string]string) PodCreateOption {
...
 }
+
+// WithDNSSearch sets the additional search domains of a container
+func WithDNSSearch(searchDomains []string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return ErrCtrFinalized
+		}
+		if StringInSlice(".", searchDomains) && len(searchDomains) > 1 {
+			return errors.Errorf("cannot pass additional search domains when also specifying '.'")
```
Comment (rhatdan) [inline libpod/options.go]: Should have a constant error here EINVAL or something, so callers to the library could check.

File: libpod/options.go [hunk 16]
```diff
@@ -636,3 +637,58 @@ func WithPodLabels(labels map[string]string) PodCreateOption {
...
+func WithDNS(dnsServers []string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return ErrCtrFinalized
+		}
+		var dns []net.IP
+		for _, i := range dnsServers {
+			result := net.ParseIP(i)
+			if result == nil {
+				return errors.Errorf("invalid IP address %s", i)
```
Comment (rhatdan) [inline libpod/options.go]: Also need EINVAL or something like that.

File: libpod/container.go [hunk 17]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+	}
+	return c.WriteStringToRundir("resolv.conf", resolv.ToString())
+}
+
+// createResolv creates a resolv struct from an input string
+func createResolv(input string) resolv {
+	var resolv resolv
+	for _, line := range strings.Split(input, "\n") {
+		if strings.HasPrefix(line, "search") {
+			fields := strings.Fields(line)
```
Comment (mheon) [inline libpod/container.go]: Might be worth checking if fields is at least 2 elements long so we don't segfault on the next line if we parse a badly formed resolv.conf? Can just ignore if the line is bad

File: libpod/container.go [hunk 18]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+			resolv.searchDomains = append(resolv.searchDomains, fields[1:]...)
+		} else if strings.HasPrefix(line, "nameserver") {
+			fields := strings.Fields(line)
```
Comment (mheon) [inline libpod/container.go]: Same as above, probably worth checking len(fields)

File: libpod/container.go [hunk 19]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+			resolv.nameServers = append(resolv.nameServers, fields[1])
+		} else if strings.HasPrefix(line, "options") {
+			fields := strings.Fields(line)
```
Comment (mheon) [inline libpod/container.go]: Same as above, probably worth checking len(fields)

File: libpod/options.go (hunk 5 above)
Comment (baude) [reply libpod/options.go]: fixed

File: libpod/container.go [hunk 20]
```diff
@@ -1590,3 +1590,113 @@ func (c *Container) Sync() error {
...
+	}
+
+	// Read and organize the hosts /etc/resolv.conf
+	resolv := createResolv(string(orig[:]))
+
+	// Populate the resolv struct with user's dns search domains
+	if len(c.config.DNSSearch) > 0 {
+		resolv.searchDomains = nil
+		// The . character means the user doesnt want any search domains in the container
+		if !StringInSlice(".", c.config.DNSSearch) {
```
Comment (mheon) [inline libpod/container.go]: I'm in favor of removing this check. We have one check already when we're adding search domains... No need to have another here.

File: libpod/options.go (hunk 7 above)
Comment (baude) [reply libpod/options.go]: this only gets called if they pass at least one in the first place

File: libpod/container.go (hunk 17 above)
Comment (baude) [reply libpod/container.go]: fixed

File: libpod/container.go (hunk 18 above)
Comment (baude) [reply libpod/container.go]: fixed

File: libpod/container.go (hunk 19 above)
Comment (baude) [reply libpod/container.go]: fixed

File: libpod/container.go (hunk 20 above)
Comment (baude) [reply libpod/container.go]: we need this check as a method to pass no search domains as discussed on irc.

File: libpod/container.go (hunk 6 above)
Comment (baude) [reply libpod/container.go]: yeah, my understanding

File: libpod/container.go (hunk 8 above)
Comment (baude) [reply libpod/container.go]: thats what "." does

File: libpod/options.go (hunk 14 above)
Comment (baude) [reply libpod/options.go]: fixed

File: libpod/container.go (hunk 13 above)
Comment (baude) [reply libpod/container.go]: fixed

File: libpod/container.go (hunk 12 above)
Comment (baude) [reply libpod/container.go]: fixed

File: libpod/container.go (hunk 11 above)
Comment (baude) [reply libpod/container.go]: fixed

File: libpod/container.go (hunk 10 above)
Comment (baude) [reply libpod/container.go]: yes

File: libpod/container.go (hunk 9 above)
Comment (baude) [reply libpod/container.go]: we need it to determine if we should should set the current data to nil (i.e. reset)

File: libpod/options.go (hunk 16 above)
Comment (baude) [reply libpod/options.go]: fixed

File: libpod/container.go (hunk 4 above)
Comment (baude) [reply libpod/container.go]: fixed

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
bot, retest this please

**Comment by mheon** [comment]
LGTM merging while everything is green

**Comment by mheon** [comment]
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
lgtm, Although we need some tests for this.

**Comment by baude** [comment]
@rhatdan looks like the tests got omitted in the rebase.  once this passes, ill open a PR for them.

**Pull request closed without merging**